cmd/gmailctl/cmd/init_cmd.go

```diff
@@ -92,23 +96,36 @@ func handleCfgDir() (err error) {
 	return createDefault(libFile, data.GmailctlLib())
 }
 
-func createDefault(path, contents string) (err error) {
+func createDefault(path, contents string) error {
 	if _, err := os.Stat(path); err != nil {
 		if !os.IsNotExist(err) {
 			return err
 		}
-		f, err := os.Create(path)
-		if err != nil {
-			return err
-		}
-		defer func() {
-			e := f.Close()
-			if err == nil {
-				err = e
-			}
-		}()
-		_, err = f.WriteString(contents)
-		return err
+		return createFile(path, contents)
 	}
 	return nil
 }
+
+func updateLib() error {
+
+	if err := handleCfgDir(); err != nil {
+		return err
+	}
+	return createFile(libFile, data.GmailctlLib())
+}
+
+func createFile(path, contents string) error {
```

Comment:
Here's the only place the named return is necessary :)

Basically the `defer` block runs after the `return`. What it does is update the error to the one returned by `Close`, but only if the error was `nil`. I hope it makes sense.

The tricky part (the one many don't like) is that you have to be careful in the body of the function not to shadow the return value variable by declaring another variable with the same name. This happens here (note the `:=`):

```
f, err := os.Create(path)
```

To avoid this, you have to use `=` and so declare both variables beforehand:

```
var f os.File
f, err = os.Create(path)
```

Now `err` is coming from the function signature `func createFile(path, contents string) (err error) {`.

Reply:
This is a very good explanation - thanks